### ISSUE TITLE
Road width

### ISSUE THREAD

**Issue by dabreegster** [jdalrym2/road_surface_classifier#1]
Hi, a colleague just pointed me to this project! I'm super excited to watch it progress. I'm curious if the current pipeline could extract the total width of roads? If this could work in urban / suburban areas and eventually match up to existing tagged center-lines, it'd be incredibly for some use cases that @robinlovelace and I have.

(One of those cases: https://github.com/a-b-street/osm2lanes, an easier way to reason about and edit lane tagging. Many roads in the US and UK and likely elsewhere have street parking, but it's rarely mapped. And even when it is, knowing the existing kerb-to-kerb width is useful for road-space reallocation planning.)

**Comment by Robinlovelace** [comment]
Agreed, this could really help out with our work and mission to generate strong evidence to support cost effective active travel interventions.

**Comment by Robinlovelace** [comment]
Great thing about osm, we could use roads that have plausible width tags on OSM as the training dataset?

**Comment by jdalrym2** [comment]
Hey there! Sorry it took me a while to get to this.

No, the current pipeline doesn't have the capability, and I fear a CNN might not be the best solution for extracting geometric information like this; at least not directly. With that said, part of the pipeline is the decoder, which is meant to replicate the mask that is fed to the network (to force the network to identify the road of interest). I'm hoping to leverage that to teach the network to identify which pixels correspond to the road as opposed to the background or any obscurations (typically trees, but also underpasses and tunnels). If the decoder works as expected, I could see extraction of geometric info from this using typical image processing techniques, given a known ground sample distance in m/px.

One thing that's related that a CNN could definitely help with your goal is identifying number of lanes and/or presence of sidewalks and street parking, as these are the typical "features" that CNNs are designed to encode. I'd love to build in something like this! This idea seems to align with the mission of osm2lanes pretty well.

Regardless, thanks for reaching out! I'll keep these thoughts in the back of my mind as this project continues to grow. Don't hesitate to contact me here or via email if you want to continue discussion related to this, or wish for updates about the project.

**Comment by Robinlovelace** [comment]
Heads-up on this @jdalrym2 my current thinking is that the segmentation step in this repo may be enough. After segmentation you can use another bit of code to compute width, e.g. by:

- Selecting one OSM way
- Sampling points along the segment
- Drawing lines perpendicular to the sampled points
- Calculating the length of the lines that intersect with pixels labelled as road (or sidewalk or other things from the segmentation stage)
- Post-process, e.g. to remove incorrect values messed-up by trees and to provide confidence intervals

Seem reasonable? See here: https://github.com/Robinlovelace/roadwidths happy to chat over this at some point? Many thanks!

**Comment by jdalrym2** [comment]
Hi @Robinlovelace, I hope you've had a wonderful New Year. Yep I think this is a wonderful idea, i.e. having a separate CV algorithm that works on the segmentation masks to extract the road widths. I imagine there could be an optimization approach here where a vector path with a certain width is fit to the image, and then something like IoU is optimized to be maximized.

I'd be happy to discuss this offline as well, email should be in my bio. Would you like some example images and segmentation masks to play with?

I'm also nearing a final model, and I'll write some code so it can be easily tied into other things. I doubt most use cases for this would be wanting to duplicate the training, but instead inference on new imagery, or expanding based on the model's output.

**Comment by Robinlovelace** [comment]
Happy to discuss but busy times for me so v. little capacity to pick up on the technical side, plus limited experience in this space. Do you know anyone who would like to pick on this? I'm wondering about doing a bit of a shout out on social media, Fosstodon and Twitter primarily, to see if we can crowd source some interest. There are some people I could ask also... Sorry for non-commital reply, glad you like the idea, let's do it!

**Comment by Robinlovelace** [comment]
Hi @jdalrym2 did you get close to a final model on this? Thanks and no worries if not, lots of options out there..

**Comment by jdalrym2** [comment]
Hey @Robinlovelace! Sorry for the delay. No I haven't gotten around to a final model on this yet. I'm happy with the model architecture and weights, but still looking into how best to inference it on real-world OSM data (e.g. you can inference multiple times on the same road in different places to build up confidence). Happy to keep you in the loop as things continue.
